src/FormAlter/MediaFormAlter.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\media_mpx\FormAlter;
+
+use Drupal\Core\DependencyInjection\ContainerInjectionInterface;
+use Drupal\Core\Entity\ContentEntityForm;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Messenger\MessengerTrait;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\media_mpx\MpxLogger;
+use Drupal\media_mpx\Repository\MpxMediaType;
+use Drupal\media_mpx\Service\UpdateVideoItem\UpdateVideoItem;
+use Drupal\media_mpx\Service\UpdateVideoItem\UpdateVideoItemRequest;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Alters the media form for mpx items to add "Reimport" button.
+ *
+ * @package Drupal\media_mpx\FormAlter
+ */
+class MediaFormAlter implements ContainerInjectionInterface {
+
+  use StringTranslationTrait;
+  use MessengerTrait;
+
+  /**
+   * The mpx media types repository.
+   *
+   * @var \Drupal\media_mpx\Repository\MpxMediaType
+   */
+  private $mpxMediaTypeRepository;
+
+  /**
+   * Update Video Item service.
+   *
+   * @var \Drupal\media_mpx\Service\UpdateVideoItem\UpdateVideoItem
+   */
+  private $updateService;
+
+  /**
+   * The custom media mpx logger service.
+   *
+   * @var \Drupal\media_mpx\MpxLogger
+   */
+  private $logger;
+
+  /**
+   * MediaFormAlter constructor.
+   */
+  private function __construct(MpxMediaType $mpxMediaTypeRepository, UpdateVideoItem $updateService, MpxLogger $logger) {
+    $this->mpxMediaTypeRepository = $mpxMediaTypeRepository;
+    $this->updateService = $updateService;
+    $this->logger = $logger;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('media_mpx.repository.mpx_media_types'),
+      $container->get('media_mpx.service.update_video_item'),
+      $container->get('media_mpx.exception_logger')
+    );
+  }
+
+  /**
+   * Implements hook_form_FORM_ID_alter().
+   */
+  public function alter(&$form, FormStateInterface $form_state, $form_id): void {
+    if ($this->alterAppliesToForm($form_state) === FALSE) {
+      return;
+    }
+
+    $form['actions']['reimport'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Reimport mpx data'),
+      '#submit' => [[$this, 'reimportCallback']],
+    ];
+  }
+
+  /**
+   * Callback for the 'reimport' button.
+   */
+  public function reimportCallback(array $form, FormStateInterface $formState) {
+    /* @var \Drupal\Core\Entity\ContentEntityForm $form_object */
+    $form_object = $formState->getFormObject();
+    $video = $form_object->getEntity();
+
+    $id_field = NULL;
+    if ($media_type = $this->mpxMediaTypeRepository->findByTypeId($video->bundle())) {
+      $field_map = $media_type->getFieldMap();
+      $id_field = $field_map['Media:id'] ?: NULL;
+    }
+
+    $mpx_id = $video->{$id_field}->value;
```

Comment:
I'd suggest being more defensive here. `$id_field` is user input. It may or may not be defined, the field may or may not be empty, and the field may or may not have a `value` property. Maybe something like:

```
try {
  if (($field = $video->get($id_field)) && !$field->isEmpty()) {
    $mpx_id = $field->getString();
  }
}
catch (\InvalidArgumentException $e) {
}
```

Reply:
Great call. In fact I'm kind of surprised $id_field can pass at null. I'm pretty sure in an earlier version of the class, I was returning early, too 🤷‍♂ . I took this chance to improve the scenario where an user hasn't configured the media source at all (e.g: items being present already from a content import from D7), and added explicit messaging about that:

![media_mpx_configure_media_source](https://user-images.githubusercontent.com/890914/58102739-1401d300-7be2-11e9-9a37-354da4b1279b.png)